File: Problem_10/AC_solution.java
import java.util.*;

class BipGraph {
    private static final int NIL = 0;
    private static final int INF = Integer.MAX_VALUE;

    private int m, n;
    private List<Integer>[] adj;
    private int[] pairU, pairV, dist;

    public BipGraph(int m, int n) {
        this.m = m;
        this.n = n;
        adj = new ArrayList[m + 1];
        for (int i = 0; i <= m; ++i) {
            adj[i] = new ArrayList<>();
        }
    }

    public void addEdge(int u, int v) {
        adj[u].add(v);
    }

    public int hopcroftKarp() {
        pairU = new int[m + 1];
        pairV = new int[n + 1];
        dist = new int[m + 1];

        Arrays.fill(pairU, NIL);
        Arrays.fill(pairV, NIL);

        int result = 0;
        while (bfs()) {
            for (int u = 1; u <= m; u++) {
                if (pairU[u] == NIL && dfs(u)) {
                    result++;
                }
            }
        }
        return result;
    }

    private boolean bfs() {
        Queue<Integer> Q = new LinkedList<>();

        for (int u = 1; u <= m; u++) {
            if (pairU[u] == NIL) {
                dist[u] = 0;
                Q.add(u);
            } else {
                dist[u] = INF;
            }
        }

        dist[NIL] = INF;

        while (!Q.isEmpty()) {
            int u = Q.poll();

            if (dist[u] < dist[NIL]) {
                for (int v : adj[u]) {
                    if (dist[pairV[v]] == INF) {
                        dist[pairV[v]] = dist[u] + 1;
                        Q.add(pairV[v]);
                    }
                }
            }
        }

        return dist[NIL] != INF;
    }

    private boolean dfs(int u) {
        if (u != NIL) {
            for (int v : adj[u]) {
                if (dist[pairV[v]] == dist[u] + 1) {
                    if (dfs(pairV[v])) {
                        pairV[v] = u;
                        pairU[u] = v;
                        return true;
                    }
                }
            }
            dist[u] = INF;
            return false;
        }
        return true;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int n = scanner.nextInt();
        int m = scanner.nextInt();

        List<List<Integer>> inj = new ArrayList<>(26);
        List<List<Integer>> don = new ArrayList<>(26);

        for (int i = 0; i < 26; i++) {
            inj.add(new ArrayList<>());
            don.add(new ArrayList<>());
        }

        for (int i = 1; i <= n; i++) {
            char x = scanner.next().charAt(0);
            inj.get(x - 'A').add(i);
        }

        for (int i = 1; i <= n; i++) {
            char x = scanner.next().charAt(0);
            don.get(x - 'A').add(i);
        }

        BipGraph g = new BipGraph(n, n);

        for (int i = 0; i < m; i++) {
            char x = scanner.next().charAt(0);
            char y = scanner.next().charAt(0);
            int ia = y - 'A';
            int ib = x - 'A';

            for (int a : inj.get(ia)) {
                for (int b : don.get(ib)) {
                    g.addEdge(a, b);
                }
            }
        }

        for (int i = 0; i < 26; i++) {
            for (int a : inj.get(i)) {
                for (int b : don.get(i)) {
                    g.addEdge(a, b);
                }
            }
        }

        if (g.hopcroftKarp() == n) {
            System.out.println("YES");
        } else {
            System.out.println("NO");
        }
    }
}

File: Problem_10/AC_Solution.py
from collections import deque


class BipGraph:
    NIL = 0
    INF = float('inf')

    def __init__(self, m, n):
        self.m = m
        self.n = n
        self.adj = [[] for _ in range(m + 1)]

    def add_edge(self, u, v):
        self.adj[u].append(v)

    def hopcroft_karp(self):
        self.pairU = [self.NIL] * (self.m + 1)
        self.pairV = [self.NIL] * (self.n + 1)
        self.dist = [0] * (self.m + 1)

        result = 0
        while self.bfs():
            for u in range(1, self.m + 1):
                if self.pairU[u] == self.NIL and self.dfs(u):
                    result += 1
        return result

    def bfs(self):
        Q = deque()

        for u in range(1, self.m + 1):
            if self.pairU[u] == self.NIL:
                self.dist[u] = 0
                Q.append(u)
            else:
                self.dist[u] = self.INF

        self.dist[self.NIL] = self.INF

        while Q:
            u = Q.popleft()

            if self.dist[u] < self.dist[self.NIL]:
                for v in self.adj[u]:
                    if self.dist[self.pairV[v]] == self.INF:
                        self.dist[self.pairV[v]] = self.dist[u] + 1
                        Q.append(self.pairV[v])

        return self.dist[self.NIL] != self.INF

    def dfs(self, u):
        if u != self.NIL:
            for v in self.adj[u]:
                if self.dist[self.pairV[v]] == self.dist[u] + 1:
                    if self.dfs(self.pairV[v]):
                        self.pairV[v] = u
                        self.pairU[u] = v
                        return True
            self.dist[u] = self.INF
            return False
        return True


if __name__ == "__main__":
    n, m = map(int, input().split())

    inj = [[] for _ in range(26)]
    don = [[] for _ in range(26)]

    x = input().split()
    for i in range(len(x)):
        inj[ord(x[i]) - ord('A')].append(i + 1)

    x = input().split()
    for i in range(len(x)):
        don[ord(x[i]) - ord('A')].append(i + 1)

    g = BipGraph(n, n)

    for i in range(m):
        x, y = input().split()
        ia = ord(y) - ord('A')
        ib = ord(x) - ord('A')

        for a in inj[ia]:
            for b in don[ib]:
                g.add_edge(a, b)

    for i in range(26):
        for a in inj[i]:
            for b in don[i]:
                g.add_edge(a, b)

    if g.hopcroft_karp() == n:
        print("YES")
    else:
        print("NO")
